File: src/vault/crypto.ts
import { createCipheriv, createDecipheriv, randomBytes, scryptSync } from 'crypto';

const ALGORITHM = 'aes-256-gcm';
const SALT_LEN = 32;
const IV_LEN = 16;
const TAG_LEN = 16;
const KEY_LEN = 32;

// scrypt parameters. Bump CURRENT_VERSION if these change.
const SCRYPT_N = 16384;
const SCRYPT_R = 8;
const SCRYPT_P = 1;

export const CURRENT_VERSION = 1;

function deriveKey(passphrase: string, salt: Buffer): Buffer {
  return scryptSync(passphrase, salt, KEY_LEN, {
    N: SCRYPT_N,
    r: SCRYPT_R,
    p: SCRYPT_P,
  });
}

export function encryptVault(plaintext: string, passphrase: string): string {
  const salt = randomBytes(SALT_LEN);
  const iv = randomBytes(IV_LEN);
  const key = deriveKey(passphrase, salt);

  const cipher = createCipheriv(ALGORITHM, key, iv);
  const ciphertext = Buffer.concat([
    cipher.update(plaintext, 'utf-8'),
    cipher.final(),
  ]);
  const tag = cipher.getAuthTag();

  return Buffer.concat([salt, iv, ciphertext, tag]).toString('base64');
}

export function decryptVault(encoded: string, passphrase: string): string {
  const buf = Buffer.from(encoded, 'base64');
  if (buf.length < SALT_LEN + IV_LEN + TAG_LEN + 1) {
    throw new Error('vault: encoded payload too short');
  }

  const salt = buf.subarray(0, SALT_LEN);
  const iv = buf.subarray(SALT_LEN, SALT_LEN + IV_LEN);
  const tag = buf.subarray(buf.length - TAG_LEN);
  const ciphertext = buf.subarray(SALT_LEN + IV_LEN, buf.length - TAG_LEN);

  const key = deriveKey(passphrase, salt);
  const decipher = createDecipheriv(ALGORITHM, key, iv);
  decipher.setAuthTag(tag);

  const plaintext = Buffer.concat([
    decipher.update(ciphertext),
    decipher.final(),
  ]);
  return plaintext.toString('utf-8');
}

File: go/internal/vault/crypto.go
package vault

import (
        "crypto/aes"
        "crypto/cipher"
        "crypto/rand"
        "encoding/base64"
        "fmt"
        "io"

        "golang.org/x/crypto/scrypt"
)

// Must match src/vault/crypto.ts exactly for cross-language vault unlock.
const (
        saltLen = 32
        ivLen   = 16
        tagLen  = 16
        keyLen  = 32
        scryptN = 16384
        scryptR = 8
        scryptP = 1
)

// Encrypt encrypts plaintext with passphrase using AES-256-GCM + scrypt.
// Wire format: base64(salt || iv || ciphertext || tag).
func Encrypt(plaintext, passphrase string) (string, error) {
        salt := make([]byte, saltLen)
        if _, err := io.ReadFull(rand.Reader, salt); err != nil {
                return "", err
        }
        iv := make([]byte, ivLen)
        if _, err := io.ReadFull(rand.Reader, iv); err != nil {
                return "", err
        }
        key, err := scrypt.Key([]byte(passphrase), salt, scryptN, scryptR, scryptP, keyLen)
        if err != nil {
                return "", err
        }
        block, err := aes.NewCipher(key)
        if err != nil {
                return "", err
        }
        gcm, err := cipher.NewGCMWithNonceSize(block, ivLen)
        if err != nil {
                return "", err
        }
        // Node's createCipheriv('aes-256-gcm', key, iv) uses iv as nonce and
        // appends auth tag separately; Go's Seal appends the tag to ciphertext.
        sealed := gcm.Seal(nil, iv, []byte(plaintext), nil)
        // sealed = ciphertext || tag
        out := make([]byte, 0, saltLen+ivLen+len(sealed))
        out = append(out, salt...)
        out = append(out, iv...)
        out = append(out, sealed...)
        return base64.StdEncoding.EncodeToString(out), nil
}

// Decrypt reverses Encrypt / Node encryptVault.
func Decrypt(encoded, passphrase string) (string, error) {
        buf, err := base64.StdEncoding.DecodeString(encoded)
        if err != nil {
                return "", fmt.Errorf("vault: base64: %w", err)
        }
        if len(buf) < saltLen+ivLen+tagLen+1 {
                return "", fmt.Errorf("vault: encoded payload too short")
        }
        salt := buf[:saltLen]
        iv := buf[saltLen : saltLen+ivLen]
        sealed := buf[saltLen+ivLen:] // ciphertext || tag
        key, err := scrypt.Key([]byte(passphrase), salt, scryptN, scryptR, scryptP, keyLen)
        if err != nil {
                return "", err
        }
        block, err := aes.NewCipher(key)
        if err != nil {
                return "", err
        }
        gcm, err := cipher.NewGCMWithNonceSize(block, ivLen)
        if err != nil {
                return "", err
        }
        plain, err := gcm.Open(nil, iv, sealed, nil)
        if err != nil {
                return "", fmt.Errorf("vault: decrypt failed (wrong passphrase or corrupt): %w", err)
        }
        return string(plain), nil
}
